packages/server/lib/cloud/artifacts/upload_artifacts.ts

```diff
@@ -0,0 +1,205 @@
+import * as errors from '../../errors'
+import exception from '../exception'
+import type { IArtifact, ArtifactUploadResult } from './artifact'
+import { createScreenshotArtifactBatch } from './screenshot_artifact'
+import { createVideoArtifact } from './video_artifact'
+import { createProtocolArtifact, composeProtocolErrorReportFromOptions } from './protocol_artifact'
+
+const debug = Debug('cypress:server:cloud:artifacts')
+
+const toUploadReportPayload = (acc: {
+  screenshots: ArtifactMetadata[]
+  video?: ArtifactMetadata
+  protocol?: ProtocolMetadata
+}, { key, ...report }: ArtifactUploadResult): UpdateInstanceArtifactsPayload => {
+  if (key === 'protocol') {
+    let { error, errorStack, allErrors } = report
+
+    if (allErrors) {
+      error = `Failed to upload Test Replay after ${allErrors.length} attempts. Errors: ${allErrors.map((error) => error.message).join(', ')}`
+      errorStack = allErrors.map((error) => error.stack).join(', ')
+    } else if (error) {
+      error = `Failed to upload Test Replay: ${error}`
+    }
+
+    debug('protocol report %O', report)
+
+    return {
+      ...acc,
+      protocol: {
+        ...report,
+        error,
+        errorStack,
+      },
+    }
+  }
+
+  return {
+    ...acc,
+    [key]: (key === 'screenshots') ? [...acc.screenshots, report] : report,
+  }
+}
+
+type UploadArtifactOptions = {
+  protocolManager: ProtocolManager
+  videoUploadUrl?: string
+  video?: string // filepath to the video artifact
+  screenshots?: {
+    screenshotId: string
+    path: string
+  }[]
+  screenshotUploadUrls?: {
+    screenshotId: string
+    uploadUrl: string
+  }[]
+  captureUploadUrl?: string
+  protocolCaptureMeta: {
+    url?: string
+    disabledMessage?: string
+  }
+  quiet?: boolean
+  runId: string
+  instanceId: string
+  spec: any
+  platform: any
+  projectId: any
+}
+
+const extractArtifactsFromOptions = async ({
+  video, videoUploadUrl, screenshots, screenshotUploadUrls, captureUploadUrl, protocolCaptureMeta, protocolManager,
+}: Pick<UploadArtifactOptions,
+  'video' | 'videoUploadUrl' |
+  'screenshots' | 'screenshotUploadUrls' |
+  'captureUploadUrl' | 'protocolManager' | 'protocolCaptureMeta'
+>): Promise<IArtifact[]> => {
+  const artifacts: IArtifact[] = []
+
+  if (videoUploadUrl && video) {
+    try {
+      artifacts.push(await createVideoArtifact(video, videoUploadUrl))
+    } catch (e) {
+      debug('Error creating video artifact: %O', e)
+    }
+  }
+
+  debug('screenshot metadata: %O', { screenshotUploadUrls, screenshots })
+  debug('found screenshot filenames: %o', screenshots)
+  if (screenshots?.length && screenshotUploadUrls?.length) {
+    const screenshotArtifacts = await createScreenshotArtifactBatch(screenshotUploadUrls, screenshots)
+
+    screenshotArtifacts.forEach((screenshot) => {
+      artifacts.push(screenshot)
+    })
+  }
+
+  try {
+    const protocolFilePath = protocolManager?.getArchivePath()
```

Comment:
```suggestion
    const protocolFilePath = protocolManager.getArchivePath()
```

Reply:
addressed in a6339bac2acb1bf0fa7fa4994aa80909b418e8ba